New Loggers

Mephi-Dross opened the pull request:
-Added support for logging into files and logging to remote servers.
-Improved structure of logging project via inheritance and interfaces.
-Added/improved tests for logging.

Commit:
Merge branch 'develop' of https://github.com/texel-sensei/ayayaa into develop

Commit:
Merge branch 'features' into develop

Commit:
feat(log): Add file logging and extensions

-Added Extension methods for the logger for ease of use.
-Added FileLogger and corresponding tests.
-Added a base Writer class. All Writers now inherit this.
This simplifies the creation of further Writers as well as
provide certain rules as to how a Writer should be structured.

Commit:
Merge branch 'develop' of https://github.com/texel-sensei/ayayaa into develop

Commit:
feat(log): Add remote logger and test

-Added writer to send messages to a remote server
-Added LoggingServer to handle messages
-Added test for LoggingServer + RemoteWriter
-Added exception variable for debugging to catch-blocks

Commit:
chore(log): Improve exception handling

-Improved exception handling by adding inner exceptions to LoggerException.

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/Writers/RemoteWriter.cs:
```diff
@@ -0,0 +1,82 @@
+                NetworkStream nwStream = client.GetStream();
+
+                // Send text...
+                Console.WriteLine("Sending : " + comparisonMessage);
```
Don't leave debug output in the code

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/Writers/RemoteWriter.cs:
```diff
@@ -0,0 +1,82 @@
+                string answer = Encoding.UTF8.GetString(bytesToRead, 0, bytesRead);
+
+                // Comparing received with sent message...
+                if (answer != comparisonMessage)
```
Remove answers to send log requests

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/Writers/ConsoleWriter.cs:
```diff
@@ -6,37 +6,33 @@
         }
 
-        private object SerializeMessage(string message)
+        protected override object SerializeMessage(string message)
```
Don't really need a SerializeMessage for the console writer

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/LoggingServer.cs:
```diff
@@ -0,0 +1,133 @@
+            loggingThread.Start();
+        }
+
+        public void StopLogging()
```
Make this class disposable to make sure that all resources are freed correctly

Commit:
chore(log): Remove server answer

-Removed the answer to the client.
-Removed message comparison in client.

Commit:
fix(log): Change IP-Address of remote tests

Commit:
Merge branch 'develop' of https://github.com/texel-sensei/ayayaa into develop

Commit:
fix(log): Make LoggingServer disposable

-Added IDisposable & corresponding function to Server.
-Implemented CancellationToken for subthread.
-Added obj/ and debug/ to gitignore.

Commit:
feat(log): Add Helper to access Logging features

-Added LoggingHelper. This class contains functions that
allow outside access to the logging system.
-Added Exception Extension for logging purposes using the
new Helper.
-Added additional test for Exception logging.

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/LoggerExtensions.cs:
```diff
@@ -0,0 +1,55 @@
+                exLogger.WriteToLogs(ex.Message, LogPriority.Error);
+                exLogger.WriteToLogs(ex.StackTrace, LogPriority.Error);
+
+                if (ex.InnerException != null)
```
Better do a loop here, in case ex.InnerException also contains an InnerException

Mephi-Dross replied:
No longer needed as the function has been removed, as it can't be used without a specific logger (e.g. exception logger).

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/LoggerExtensions.cs:
```diff
@@ -0,0 +1,55 @@
+
+        public static void LogThis(this Exception ex)
+        {
+            Logger exLogger = LoggingHelper.FindLogger("EX");
```
Logger "classes" should not be one logger for exceptions, one logger for warnings etc, but based on module/subsystem. Have one logger for the database connection, one logger for the web API, etc

Mephi-Dross replied:
Has been removed. Necessary Loggers can be added in the LoggingHelper either in the CreateDefault or just at runtime during initialisation.

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/Writers/RemoteWriter.cs:
```diff
@@ -0,0 +1,81 @@
+
+                // Receive answer...
+                //byte[] bytesToRead = new byte[client.ReceiveBufferSize];
+                //int bytesRead = nwStream.Read(bytesToRead, 0, client.ReceiveBufferSize);
```
No commented out code!

texel-sensei commented on src/core/ayayaa-server/tests/logging/LoggingTest.cs:
```diff
@@ -68,4 +93,115 @@ private void WriteToLogs()
+                LogPriority priority5 = LogPriority.Error;
+                LogPriority priority6 = LogPriority.FIRE;
+
+                server.LocalLogger.AddWriter(new FileWriter("E:\\Programming\\Tests and Temp\\ServerLog1"), LogPriority.Trace);
```
I don't have an E:\ drive, will the test fail for me?
What's about linux without drive letters?
Get a temporary path instead: https://docs.microsoft.com/en-us/dotnet/api/system.io.path.gettemppath?view=netframework-4.7.2

Mephi-Dross replied:
Has been changed to temp path.

texel-sensei replied:
I just noticed, that this function does not what I expected it to do (returns the directory behind %TMP% and does not return a unique name). But that is a better implemented elsewhere and just keep it like that for now

Commit:
refactor(log): Adjust comments

-Removed unnecessary comment informations.
-Removed ConsoleWriter functions.
-Removed ExceptionLogger and corresponding code.

texel-sensei commented on src/core/ayayaa-server/ayayaa-logging/LoggerExtensions.cs:
```diff
@@ -36,20 +36,5 @@ public static void LogFIRE(this Logger logger, string message)
         {
             logger.WriteToLogs(message, LogPriority.FIRE);
         }
-
```
I didn't mean to remove this code, keep it as an extension method on the logger, with the exception as parameter.
Logging exceptions + inner exceptions + full stack trace is really handy.

Commit:
Merge branch 'develop' into logging_feature_IN#7

Commit:
chore(log): Fix branch issues

-Fixed missing files because of project changes.

Commit:
fix(log): Adjust exception handling

-Added handling for specific SocketException 10004.
-Adjusted tests to properly test dispose.

Commit:
fix(log): Adjust thread handling

-Made IsRunning volatile.
-Use IsRunning instead of CancellationToken.
-Does it work? Hell if I know -.-

texel-sensei commented on ayayaa-server/ayayaa-logging/LoggingServer.cs:
```diff
@@ -0,0 +1,177 @@
+        }
+
+
+        public Logger LocalLogger { get; set; }
```
Public access to a variable that is used in a private thread.
Race condition waiting to happen.
